src/core/str_hash_table.c

```diff
@@ -0,0 +1,494 @@
+                    && MVM_string_substrings_equal_nocheck(tc, key, 0,
+                                                           MVM_string_graphs_nocheck(tc, key),
+                                                           entry->key, 0))) {
+                return entry;
+            }
+        }
+        ++probe_distance;
+        ++metadata;
+        entry_raw += hashtable->entry_size;
+        assert(probe_distance <= MVM_HASH_MAX_PROBE_DISTANCE);
+        assert(metadata < hashtable->metadata + hashtable->official_size + hashtable->max_items);
+        assert(metadata < hashtable->metadata + hashtable->official_size + 256);
+    }
+}
+
+void *MVM_str_hash_lvalue_fetch_nt(MVMThreadContext *tc,
+                                   MVMStrHashTable *hashtable,
+                                   MVMString *key) {
+    if (MVM_UNLIKELY(hashtable->entries == NULL)) {
+        MVM_str_hash_initial_allocate(tc, hashtable, 0);
+    }
+    else if (MVM_UNLIKELY(hashtable->cur_items >= hashtable->max_items)) {
+        /* We should avoid growing the hash if we don't need to.
+         * It's expensive, and for hashes with iterators, growing the hash
+         * invalidates iterators. Which is buggy behaviour if the fetch doesn't
+         * need to create a key. */
+        struct MVMStrHashHandle *entry = MVM_str_hash_fetch_nt(tc, hashtable, key);
+        if (entry) {
+            return entry;
+        }
+
+        MVMuint32 true_size =  hash_true_size(hashtable);
+        char *entry_raw_orig = hashtable->entries;
+        MVMuint8 *metadata_orig = hashtable->metadata;
+
+        hash_grow(tc, hashtable);
+
+        char *entry_raw = entry_raw_orig;
+        MVMuint8 *metadata = metadata_orig;
+        MVMHashNumItems bucket = 0;
+        while (bucket < true_size) {
+            if (*metadata) {
+                struct MVMStrHashHandle *old_entry = (struct MVMStrHashHandle *) entry_raw;
+                void *new_entry_raw = hash_insert_internal(tc, hashtable, old_entry->key);
+                struct MVMStrHashHandle *new_entry = (struct MVMStrHashHandle *) new_entry_raw;
+                assert(new_entry->key == NULL);
+                memcpy(new_entry, old_entry, hashtable->entry_size);
+            }
+            ++bucket;
+            ++metadata;
+            entry_raw += hashtable->entry_size;
+        }
+        MVM_free(entry_raw_orig);
+        MVM_free(metadata_orig - 1);
+    }
+    struct MVMStrHashHandle *new_entry
+        = hash_insert_internal(tc, hashtable, key);
+    if (!new_entry->key) {
+        ++hashtable->cur_items;
+    }
+    return new_entry;
+}
+
+/* UNCONDITIONALLY creates a new hash entry with the given key and value.
+ * Doesn't check if the key already exists. Use with care.
+ * (well that's the official line. As you can see, the XXX suggests we currently
+ * don't exploit the documented freedom.) */
+void *MVM_str_hash_insert_nt(MVMThreadContext *tc,
```

Comment:
I don't see the `XXX` this refers to.

Reply:
What does `nt` mean here? At first I thought "null terminated" (we use it for that in some of the string handling stuff) and got worried, but it doesn't seem to mean that here...

Reply:
`nt` - *no throw*. I don't **think** that I was the first person to add this to the codebase. But I haven't checked. Better name would be fine, but I don't have a great suggestion.

Reply:
We use `nocheck` in quite a few places for "doesn't check stuff and just goes ahead and does the operation". May be applicable here.

Reply:
Oh bother. Yes, the `XXX` got refactored out. It used to be a comment about roughly "OK, we said that we didn't check for duplicates, but we just did, and now what should we do?". It's now that "oops", which (arguably) still contradicts the official documentation of "UNCONDITIONALLY". "oops and exit" is not the documented interface (of "la la la, I can't hear you")